File: app/src/main/java/niks/foreignreader/ReaderActivity/ReaderPageFragment.java
package niks.foreignreader.ReaderActivity;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Spannable;
import android.text.SpannableString;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.BreakIterator;
import java.util.Locale;

import niks.foreignreader.LongClickLinkMovementMethod;
import niks.foreignreader.R;

/**
 * Created by Niks on 28.03.2017.
 */

public class ReaderPageFragment extends Fragment {

    static final String ARGUMENT_PAGE_NUMBER = "arg_page_number";
    static final String ARGUMENT_PAGE_TEXT = "arg_page_text";
    private ReaderClickableSpan mReaderClickableSpan;
    private View.OnLongClickListener mTextViewOnLongClickListener;
    int pageNumber;
    String pageText;

    {
        mReaderClickableSpan = new ReaderClickableSpan() {

            @Override
            public void onClick(View v) {
            }

        };
    }

    public void setTextViewOnLongClickListener(View.OnLongClickListener textViewOnLongClickListener) {
        mTextViewOnLongClickListener = textViewOnLongClickListener;
    }

    public void setReaderClickableSpan(ReaderClickableSpan readerClickableSpan) {
        mReaderClickableSpan = readerClickableSpan;
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        pageNumber = getArguments().getInt(ARGUMENT_PAGE_NUMBER);
        pageText = getArguments().getString(ARGUMENT_PAGE_TEXT);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_reader, null);

        TextView textViewFragmentReader = (TextView) view.findViewById(R.id.textViewFragmentReader);

        textViewFragmentReader.setMovementMethod(LongClickLinkMovementMethod.getInstance());
        if (mTextViewOnLongClickListener != null) {
            textViewFragmentReader.setOnLongClickListener(mTextViewOnLongClickListener);
        }

        textViewFragmentReader.setText(getSpannableText(pageText), TextView.BufferType.SPANNABLE);
        return view;
    }

    private Spannable getSpannableText(String text) {
        final String trimedText = text.trim();
        Spannable spans = new SpannableString(trimedText);
        BreakIterator iterator = BreakIterator.getWordInstance(Locale.US);
        iterator.setText(trimedText);
        int start = iterator.first();
        for (int end = iterator.next(); end != BreakIterator.DONE; start = end, end = iterator
                .next()) {
            String possibleWord = trimedText.substring(start, end);
            if (Character.isLetterOrDigit(possibleWord.charAt(0))) {
                ReaderClickableSpan clickSpan = mReaderClickableSpan.setWord(possibleWord).clone();
                //mReaderClickableSpan.setWord(possibleWord);
                spans.setSpan(clickSpan, start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        }
        return spans;
    }
}

File: app/src/main/java/niks/foreignreader/presentation/reader/ReaderPageFragment.kt
package niks.foreignreader.presentation.reader

import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.Spannable
import android.text.SpannableString
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import kotlinx.android.synthetic.main.item_reader_page.*
import niks.foreignreader.R
import java.text.BreakIterator
import java.util.*

class ReaderPageFragment : Fragment() {

    private var pageNumber: Int? = null
    private var content: String? = null
    private var listener: ReaderListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.item_reader_page, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        textViewFragmentReader.movementMethod = LongClickLinkMovementMethod.getInstance()
        if (listener != null) {
            textViewFragmentReader.setOnLongClickListener(listener)
        }

        content?.let {
            textViewFragmentReader.setText(getSpannable(it), TextView.BufferType.SPANNABLE)
        }
    }

    fun setListener(listener: ReaderListener) {
        this.listener = listener
    }

    fun setContent(pageNumber: Int, content: String) {
        this.content = content
        this.pageNumber = pageNumber

        textViewFragmentReader?.setText(getSpannable(content), TextView.BufferType.SPANNABLE)
    }

    private fun getSpannable(text: String): Spannable {
        val trimText = text.trim { it <= ' ' }
        val spans = SpannableString(trimText)
        val iterator = BreakIterator.getWordInstance(Locale.US)
        iterator.setText(trimText)
        var start = iterator.first()
        var end = iterator.next()
        while (end != BreakIterator.DONE) {
            val possibleWord = trimText.substring(start, end)
            if (Character.isLetterOrDigit(possibleWord[0])) {
                val clickSpan = object : ReaderClickableSpan() {
                    override fun onClick(widget: View) {
                        listener?.onClickSpan(text)
//                        this@ReaderFragment.sendYandexTranslateQuery(text)
                    }
                }
                clickSpan.setWord(possibleWord)
                spans.setSpan(clickSpan, start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            start = end
            end = iterator.next()
        }
        return spans
    }
}
